### PR TITLE
Be-9-add-company-actions

### PR DESCRIPTION
**Pull request by Randayy**

### CODE REVIEW

File: app/db/user_models.py [hunk 1]
```diff
@@ -1,15 +1,86 @@
...
+    owner_id = Column(UUID(as_uuid=True), ForeignKey(
+        'users.id'), default=uuid.uuid4, nullable=False)
+    owner = relationship('User', back_populates='companies')
+    members = relationship('Company_Members', back_populates='company')
+    invitations = relationship('Invitations', back_populates='company')
+    requests = relationship('Requests', back_populates='company')
+    visible = Column(Boolean, default=True, nullable=False)
+
+
+class Company_Members(BaseTable):
```
Comment (Artemooon) [inline app/db/user_models.py]: change the name - CompanyMember

File: app/db/user_models.py [hunk 2]
```diff
@@ -1,15 +1,86 @@
...
+    __tablename__ = 'company_members'
+
+    company_id = Column(UUID(as_uuid=True), ForeignKey(
+        'company.id'), nullable=False)
+    user_id = Column(UUID(as_uuid=True), ForeignKey(
+        'users.id'), nullable=False)
+    company = relationship('Company', back_populates='members')
```
Comment (Artemooon) [inline app/db/user_models.py]: Add member role field

File: app/db/user_models.py [hunk 3]
```diff
@@ -1,15 +1,86 @@
...
+
+
+class Requests(BaseTable):
+    __tablename__ = 'requests'
+
+    status = Column(EnumColumn(RequestsStatus), nullable=False)
+    company_id = Column(UUID(as_uuid=True), ForeignKey('company.id'), nullable=False)
+    user_id = Column(UUID(as_uuid=True), ForeignKey('users.id'), nullable=False)
+    user = relationship('User', back_populates='requests')
+    company = relationship('Company', back_populates='requests')
```
Comment (Artemooon) [inline app/db/user_models.py]: It's better to use a single model for the company Action model, just add status and use such options:
requested,
invited,
kicked,
accepted

File: app/repositories/company_repository.py [hunk 4]
```diff
@@ -68,5 +72,178 @@ async def update_company(self, company_id: UUID, company_data: dict) -> Company:
...
+
+    async def get_invitation(self, company_id: UUID, current_user: User):
+        invitation = await self.db.execute(select(Invitations).where(Invitations.company_id == company_id).where(Invitations.user_id == current_user.id))
+        invitation = invitation.scalars().first()
+        if not invitation:
+            raise HTTPException(status_code=404, detail="Invitation not found")
+        return invitation
+
+    async def cancel_invitation(self, company_id: UUID, user_id: UUID) -> None:
+        invitation = await self.get_invitation(company_id, user_id)
```
Comment (Artemooon) [inline app/repositories/company_repository.py]: it's better to save the info that invitation was cancelled, do not remove, but change the status

File: app/routers/company_routes.py [hunk 5]
```diff
@@ -43,3 +43,91 @@ async def update_company(company_id: UUID, company_data: CompanyCreateSchema, db
...
+
+
+@company_router.delete("/cancel_join_request/{company_id}")
+async def cancel_join_request(company_id: UUID, db: AsyncSession = Depends(get_session), current_user: User = Depends(get_current_user_from_token)):
+    service = CompanyService(db)
+    await service.cancel_join_request(company_id, current_user)
+    return {"message": "Join request cancelled successfully"}
+
+
+@company_router.post("/accept_join_request/{company_id}/{user_id}")
```
Comment (Artemooon) [inline app/routers/company_routes.py]: use - instead of _ 
/accept-join-request
use the same for all endpoints

File: app/db/user_models.py (hunk 1 above)
Comment (Randayy) [reply app/db/user_models.py]: changed

File: app/db/user_models.py (hunk 2 above)
Comment (Randayy) [reply app/db/user_models.py]: added in task 10

File: app/repositories/company_repository.py (hunk 4 above)
Comment (Randayy) [reply app/repositories/company_repository.py]: saved the canceled invitation

File: app/routers/company_routes.py (hunk 5 above)
Comment (Randayy) [reply app/routers/company_routes.py]: fixed

File: app/db/user_models.py (hunk 3 above)
Comment (Randayy) [reply app/db/user_models.py]: fixed check changes

File: app/routers/company_routes.py [hunk 6]
```diff
@@ -43,3 +43,91 @@ async def update_company(company_id: UUID, company_data: CompanyCreateSchema, db
...
+    service = CompanyService(db)
+    await service.cancel_join_request(company_id, current_user)
+    return {"message": "Join request cancelled successfully"}
+
+
+@company_router.post("/accept-join-request/{company_id}/{user_id}")
+async def accept_join_request(company_id: UUID, user_id: UUID, db: AsyncSession = Depends(get_session), current_user: User = Depends(get_current_user_from_token)):
+    service = CompanyService(db)
+    await service.accept_join_request(company_id, user_id, current_user)
+    return {"message": "Join request accepted successfully"}
```
Comment (Artemooon) [inline app/routers/company_routes.py]: create Pydantic model with the message, you do not return status code

File: app/routers/company_routes.py [hunk 7]
```diff
@@ -43,3 +43,91 @@ async def update_company(company_id: UUID, company_data: CompanyCreateSchema, db
...
+
+
+@company_router.post("/reject-join-request/{company_id}/{user_id}")
+async def reject_join_request(company_id: UUID, user_id: UUID, db: AsyncSession = Depends(get_session), current_user: User = Depends(get_current_user_from_token)):
+    service = CompanyService(db)
+    await service.reject_join_request(company_id, user_id, current_user)
+    return {"message": "Join request rejected successfully"}
```
Comment (Artemooon) [inline app/routers/company_routes.py]: create Pydantic model with the message, you do not return status code

Comment (Randayy) [reply app/routers/company_routes.py]: in each route I should return pydantic model ?

File: app/repositories/company_repository.py [hunk 8]
```diff
@@ -68,5 +72,182 @@ async def update_company(self, company_id: UUID, company_data: dict) -> Company:
...
+        if invitation:
+            raise HTTPException(status_code=404, detail="User already invited")
+        return None
+
+    async def check_if_user_requested_already(self, company_id: UUID, user_id: UUID) -> None:
+        action = await self.db.execute(select(Action).where(Action.company_id == company_id).where(Action.user_id == user_id).where(Action.status == ActionStatus.REQUESTED))
+        action = action.scalars().first()
+        if action.status == ActionStatus.REQUESTED:
+            raise HTTPException(
+                status_code=404, detail="User already requested")
```
Comment (Artemooon) [inline app/repositories/company_repository.py]: 404 bad status code

File: app/repositories/company_repository.py [hunk 9]
```diff
@@ -68,5 +72,182 @@ async def update_company(self, company_id: UUID, company_data: dict) -> Company:
...
+        if company_member:
+            raise HTTPException(
+                status_code=404, detail="User already member of company")
+        return None
+
+    async def check_if_user_invited_already(self, company_id: UUID, user_id: UUID) -> None:
+        invitation = await self.db.execute(select(Action).where(Action.company_id == company_id).where(Action.user_id == user_id).where(Action.status == ActionStatus.INVITED))
+        invitation = invitation.scalars().first()
+        if invitation:
+            raise HTTPException(status_code=404, detail="User already invited")
```
Comment (Artemooon) [inline app/repositories/company_repository.py]: 404 bad status code

File: app/repositories/company_repository.py [hunk 10]
```diff
@@ -68,5 +72,182 @@ async def update_company(self, company_id: UUID, company_data: dict) -> Company:
...
+    async def reject_invitation(self, company_id: UUID, current_user: User) -> None:
+        action = await self.get_invitation(company_id, current_user)
+        if action.status == ActionStatus.INVITED:
+            action.status = ActionStatus.REJECTED
+            await self.db.commit()
+            logging.info("Invitation rejected")
+            return None
+        else:
+            raise HTTPException(
+                status_code=404, detail="Invitation already accepted")
```
Comment (Artemooon) [inline app/repositories/company_repository.py]: use another status code here

File: app/routers/company_routes.py (hunk 7 above)
Comment (Artemooon) [reply app/routers/company_routes.py]: yes

File: app/routers/user_routes.py [hunk 11]
```diff
@@ -64,3 +64,17 @@ async def login_for_token(form_data: Annotated[OAuth2PasswordRequestForm, Depend
...
+
+
+@user_router.get("/my_invitations", response_model=UserInvitationListSchema)
+async def get_my_invitations(db: AsyncSession = Depends(get_session), current_user: User = Depends(get_current_user_from_token)):
+    service = UserService(db)
+    invitations = await service.get_my_invitations(current_user)
+    return invitations
+
+
+@user_router.post("/my_requests")
```
Comment (Artemooon) [inline app/routers/user_routes.py]: my-requests

File: app/routers/user_routes.py [hunk 12]
```diff
@@ -64,3 +64,17 @@ async def login_for_token(form_data: Annotated[OAuth2PasswordRequestForm, Depend
 @user_router.get("/me", response_model=UserDetailSchema)
 async def read_users_me(db: AsyncSession = Depends(get_session), current_user: User = Depends(get_current_user_from_token)):
     return UserDetailSchema(**current_user.__dict__)
+
+
+@user_router.get("/my_invitations", response_model=UserInvitationListSchema)
```
Comment (Artemooon) [inline app/routers/user_routes.py]: my-invitations

File: app/services/company_services.py [hunk 13]
```diff
@@ -46,3 +46,96 @@ async def check_if_owner_of_company(self, company_id: UUID, current_user: User):
...
+        await self.check_if_owner_of_company(company_id, current_user)
+        await self.check_if_user_is_member_of_company_for_deleting(company_id, user_id)
+        await self.check_if_owner_not_take_himself(current_user, user_id)
+        await self.company_repository.delete_user_from_company(company_id, user_id)
+        return {"message": "User deleted successfully"}
+
+    async def exit_from_company(self, company_id: UUID, current_user: User):
+        await self.check_if_user_is_member_of_company_for_deleting(company_id, current_user.id)
+        await self.company_repository.exit_from_company(company_id, current_user)
+        return {"message": "You exited from company successfully"}
```
Comment (Artemooon) [inline app/services/company_services.py]: you do not use this dict

File: app/services/company_services.py [hunk 14]
```diff
@@ -46,3 +46,96 @@ async def check_if_owner_of_company(self, company_id: UUID, current_user: User):
...
+    async def reject_invitation(self, company_id: UUID, current_user: User):
+        await self.company_repository.reject_invitation(company_id, current_user)
+        return {"message": "Invitation rejected successfully"}
+
+    async def delete_user_from_company(self, company_id: UUID, user_id: UUID, current_user: User):
+        await self.check_if_owner_of_company(company_id, current_user)
+        await self.check_if_user_is_member_of_company_for_deleting(company_id, user_id)
+        await self.check_if_owner_not_take_himself(current_user, user_id)
+        await self.company_repository.delete_user_from_company(company_id, user_id)
+        return {"message": "User deleted successfully"}
```
Comment (Artemooon) [inline app/services/company_services.py]: you do not use this dict

File: app/services/company_services.py [hunk 15]
```diff
@@ -46,3 +46,96 @@ async def check_if_owner_of_company(self, company_id: UUID, current_user: User):
...
+        await self.check_if_user_is_member_of_company(company_id, user_id)
+        await self.check_if_user_invited_already(company_id, user_id)
+        await self.check_if_owner_not_take_himself(current_user, user_id)
+        await self.company_repository.invite_user_to_company(company_id, user_id)
+        return {"message": "User invited successfully"}
+
+    async def cancel_invitation(self, company_id: UUID, user_id: UUID, current_user: User):
+        await self.check_if_owner_of_company(company_id, current_user)
+        await self.company_repository.cancel_invitation(company_id, user_id)
+        return {"message": "Invitation cancelled successfully"}
```
Comment (Artemooon) [inline app/services/company_services.py]: you do not use this dict

File: app/services/company_services.py [hunk 16]
```diff
@@ -46,3 +46,96 @@ async def check_if_owner_of_company(self, company_id: UUID, current_user: User):
...
+        await self.check_if_owner_not_request(company_id, current_user.id)
+        await self.check_if_user_is_member_of_company(company_id, current_user.id)
+        await self.check_if_user_invited_already(company_id, current_user.id)
+        await self.check_if_user_requested_already(company_id, current_user.id)
+        await self.company_repository.send_join_request(company_id, current_user.id)
+        return None
+
+    async def cancel_join_request(self, company_id: UUID, current_user: User) -> None:
+        await self.company_repository.cancel_join_request(company_id, current_user.id)
+        return None
```
Comment (Artemooon) [inline app/services/company_services.py]: remove return None everywhere 
function without return - returns None by default

File: app/routers/user_routes.py (hunk 12 above)
Comment (Randayy) [reply app/routers/user_routes.py]: fixed

File: app/routers/user_routes.py (hunk 11 above)
Comment (Randayy) [reply app/routers/user_routes.py]: fixed

File: app/services/company_services.py (hunk 16 above)
Comment (Randayy) [reply app/services/company_services.py]: deleted return none

File: app/services/company_services.py (hunk 15 above)
Comment (Randayy) [reply app/services/company_services.py]: deleted

File: app/services/company_services.py (hunk 14 above)
Comment (Randayy) [reply app/services/company_services.py]: deleted

File: app/services/company_services.py (hunk 13 above)
Comment (Randayy) [reply app/services/company_services.py]: deleted

File: app/repositories/company_repository.py (hunk 8 above)
Comment (Randayy) [reply app/repositories/company_repository.py]: fixed to 409

File: app/repositories/company_repository.py (hunk 9 above)
Comment (Randayy) [reply app/repositories/company_repository.py]: fixed to 409 fixed

File: app/repositories/company_repository.py (hunk 10 above)
Comment (Randayy) [reply app/repositories/company_repository.py]: used 409

File: app/routers/company_routes.py (hunk 6 above)
Comment (Randayy) [reply app/routers/company_routes.py]: created

File: app/routers/company_routes.py (hunk 7 above)
Comment (Randayy) [reply app/routers/company_routes.py]: created

**Comment by Randayy** [comment]
created pedantic schema to return message with company action, added cascade deleting, I tested and if you delete user , company member also is deleted ,actions with this user also are deleted and if you delete company , all actions are deleted and all company members are deleted, but users saved

**Comment by Randayy** [comment]
fixed comments

**Pull request merged**